### ISSUE TITLE
Inconsistent comparison behavior in combination with OR

### ISSUE THREAD

**Issue by pfu3tz** [arangodb/arangodb#13655]
## My Environment

* __ArangoDB Version__:        3.7.8
* __Storage Engine__:            RocksDB 6.8.0
* __Deployment Mode__:      Single Server
* __Deployment Strategy__:   Manual Start
* __Configuration__:               -
* __Infrastructure__:               Own
* __Operating System__:        Ubuntu 18.04.5 LTS
* __Total RAM in your machine__:        16 GB
* __Disks in use__:       SSD
* __Used Package__:         Ubuntu .deb

## Component, Query & Data

__Affected feature__:
arangosh, but was originally detected with Java driver version 6.9.0

__AQL query (if applicable)__:
The affected query as I understand the documentation:
```
FOR doc IN table FILTER ((true >= doc.column) OR doc.column) RETURN doc
```
__AQL explain and/or profile (if applicable)__:
Query String (71 chars, cacheable: true):
 FOR doc IN table FILTER ((true >= doc.column) OR doc.column) RETURN doc

Execution plan:
 Id   NodeType                  Est.   Comment
  1   SingletonNode                1   * ROOT
  2   EnumerateCollectionNode      1     - FOR doc IN table   /* full collection scan */   FILTER (doc.`column` <= true)   /* early pruning */
  5   ReturnNode                   1       - RETURN doc

Indexes used:
 none

Optimization rules applied:
 Id   RuleName
  1   remove-redundant-or
  2   move-filters-into-enumerate

__Dataset__:
// This is what I ran:

// DB Creation and Insertion
db._dropDatabase("test")
db._createDatabase("test")
db._useDatabase("test")

db._create("table")
db._query("INSERT {column: 1} IN table")

// Without the Filter 
db._query("FOR doc IN table RETURN doc")
// db._explain("FOR doc IN table RETURN doc")

// The Query
db._query("FOR doc IN table FILTER (true >= doc.column) OR doc.column RETURN doc")
// db._explain("FOR doc IN table FILTER (true >= doc.column) OR doc.column RETURN doc")

// The negated query, in case I misunderstood how the comparison is evaluated
db._query("FOR doc IN table FILTER !(((true >= doc.column) OR doc.column)) RETURN doc")
// db._explain("FOR doc IN table FILTER !((true >= doc.column) OR doc.column) RETURN doc")

## Steps to reproduce
1. Run the script
2. Check if the single document shows up in either the negated or the query

__Problem__:
As I read the documentation: true >= doc.column should evaluate to false, because comparisons do not make type conversions and return false in that case: see [here](https://www.arangodb.com/docs/stable/aql/operators.html#comparison-operators). That means that the OR evaluates to true because doc.column is a number type not equal to 0 and the document should be returned.

In any case I also added the negation, because I could have misread the documentation but this does also not return the document.

**Comment by Simran-B** [comment]
I agree that the following statement makes it sound like comparisons would return false if their types mismatch:

> The comparison operators will not perform any implicit type casts if the compared operands have different or non-sensible types.

This is contradicted by https://www.arangodb.com/docs/stable/aql/fundamentals-type-value-order.html

> When checking for equality or inequality or when determining the sort order of values, AQL uses a deterministic algorithm that takes both the data types and the actual values into account.
>
> The compared operands are first compared by their data types, and only by their data values if the operands have the same data types.

The former statement seems to be wrong or is at least phrased confusingly. I'll try to find out what this was intended to express.

**Comment by pfu3tz** [comment]
Thank you for pointing that out, I agree and read through the fundamentals and am still convinced there is something else here.

First of all I should have noted that no matter what the comparison evaluates to, it should return the document, because it ors with a true statement: doc.column is not equals zero and therefore should evaluated to true for that document. Additionally, if that were not the case the negation should return the document as however the boolean is finally evaluated one of the two has to.

But I did some more testing and I found a better argument than just negating the query:
`db._query("FOR doc IN table FILTER (true >= doc.column) OR doc.column RETURN doc")`
This is what I wrote down as the query, now this:
`db._query("FOR doc IN table FILTER doc.column OR (true >= doc.column) RETURN doc")`
This is the same query but we flip the or statement, which should make no difference. But this actually returns the document while the first one does not.

**Comment by Simran-B** [comment]
This looks like a bug indeed. The conditions should be commutative. I created an internal ticket BTS-340.

As a workaround, you may disable the `remove-redundant-or` optimization:

```js
db._query(`FOR val IN [false, true, 0, 1] FILTER val <= true || val RETURN val`,
  null,
  { optimizer: { rules: ["-remove-redundant-or"] } }
);
```

**Comment by Simran-B** [comment]
Regarding this statement:
> The comparison operators will not perform any implicit type casts if the compared operands have different or non-sensible types.

It was apparently intended to refer to the operators `==`, `!=` and IN only. `1` is different to `"1"`, `[1]`, `["1"]` and `true` for instance, whereas in JavaScript `1 == "1" == [1] == ["1"] == true` evaluates to true. AQL checks for strict equality.

**Comment by goedderz** [comment]
This should be fixed in 3.7.11 via https://github.com/arangodb/arangodb/pull/14035.

**Comment by maxkernbach** [comment]
Hi @pfu3tz,

Version 3.7.11 including a fix has been released.

Could you please let us know whether the issue is resolved?

**Comment by pfu3tz** [comment]
Yes, has been resolved, thank you all.
